conn.go

```diff
@@ -418,69 +422,87 @@ func (c *Conn) handleStartTLS() {
 // DATA
 func (c *Conn) handleData(arg string) {
 	if arg != "" {
-		c.WriteResponse(501, "DATA command should not have any arguments")
+		c.WriteResponse(501, [3]int{5, 5, 4}, "DATA command should not have any arguments")
 		return
 	}
 
 	if !c.fromReceived || len(c.recipients) == 0 {
-		c.WriteResponse(502, "Missing RCPT TO command.")
+		c.WriteResponse(502, [3]int{5, 5, 1}, "Missing RCPT TO command.")
 		return
 	}
 
 	// We have recipients, go to accept data
-	c.WriteResponse(354, "Go ahead. End your data with <CR><LF>.<CR><LF>")
+	c.WriteResponse(354, [3]int{2, 0, 0}, "Go ahead. End your data with <CR><LF>.<CR><LF>")
 
 	var (
-		code int
-		msg  string
+		code         int
+		enhancedCode [3]int
+		msg          string
 	)
 	r := newDataReader(c)
 	err := c.Session().Data(r)
 	io.Copy(ioutil.Discard, r) // Make sure all the data has been consumed
 	if err != nil {
 		if smtperr, ok := err.(*SMTPError); ok {
 			code = smtperr.Code
+			enhancedCode = smtperr.EnhancedCode
 			msg = smtperr.Message
 		} else {
 			code = 554
+			enhancedCode = [3]int{5, 0, 0}
 			msg = "Error: transaction failed, blame it on the weather: " + err.Error()
 		}
 	} else {
 		code = 250
+		enhancedCode = [3]int{2, 0, 0}
 		msg = "OK: queued"
 	}
 
 	if c.server.LMTP {
 		// TODO: support per-recipient responses
 		for _, rcpt := range c.recipients {
-			c.WriteResponse(code, "<"+rcpt+"> "+msg)
+			c.WriteResponse(code, enhancedCode, "<"+rcpt+"> "+msg)
 		}
 	} else {
-		c.WriteResponse(code, msg)
+		c.WriteResponse(code, enhancedCode, msg)
 	}
 
 	c.resetMessage()
 }
 
 func (c *Conn) Reject() {
-	c.WriteResponse(421, "Too busy. Try again later.")
+	c.WriteResponse(421, [3]int{4, 4, 5}, "Too busy. Try again later.")
 	c.Close()
 }
 
 func (c *Conn) greet() {
-	c.WriteResponse(220, fmt.Sprintf("%v ESMTP Service Ready", c.server.Domain))
+	c.WriteResponse(220, NoEnhancedCode, fmt.Sprintf("%v ESMTP Service Ready", c.server.Domain))
 }
 
-func (c *Conn) WriteResponse(code int, text ...string) {
+func (c *Conn) WriteResponse(code int, enhCode [3]int, text ...string) {
 	// TODO: error handling
 	if c.server.WriteTimeout != 0 {
 		c.conn.SetWriteDeadline(time.Now().Add(c.server.WriteTimeout))
 	}
 
+	// All responses must include an enhanced code, if it is missing - use
+	// a generic code X.0.0.
+	if enhCode == EnhancedCodeNotSet {
+		cat := code / 100
+		switch cat {
+		case 2, 4, 5:
```

Comment:
I this doesn't happen, `enhCode` will still be set to `EnhancedCodeNotSet` and will be sent as-is on the wire. I guess this isn't desirable?

Reply:
Oops.

Reply:
5be3db175beba591fd32a49ee48465fa30415448